As a deeper issue I just discovered that fabric packs have likely been failing to launch on windows forever because of this.

```
Exception caught from launcher
java.lang.NoClassDefFoundError: org/spongepowered/asm/launch/MixinBootstrap
	at net.fabricmc.loader.impl.game.LoaderLibrary.<clinit>(LoaderLibrary.java:41)
	at net.fabricmc.loader.impl.game.LibClassifier.<init>(LibClassifier.java:100)
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.locateGame(MinecraftGameProvider.java:173)
	at net.fabricmc.loader.impl.launch.knot.Knot.createGameProvider(Knot.java:175)
	at net.fabricmc.loader.impl.launch.knot.Knot.init(Knot.java:130)
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:68)
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23)
	at org.prismlauncher.launcher.impl.StandardLauncher.launch(StandardLauncher.java:100)
	at org.prismlauncher.EntryPoint.listen(EntryPoint.java:129)
	at org.prismlauncher.EntryPoint.main(EntryPoint.java:70)
Caused by: java.lang.ClassNotFoundException: org.spongepowered.asm.launch.MixinBootstrap
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(Unknown Source)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(Unknown Source)
	at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
	... 10 more
Exiting with ERROR
[21:59:49] [WARN] [FabricLoader/Knot]: Class path entries reference missing files: C:/Users/ryexc/AppData/Roaming/PrismLauncher/libraries/net/fabricmc/sponge-mixin/0.13.3+mixin.0.8.5/sponge-mixin-0.13.3+mixin.0.8.5.jar - the game may not load properly!
```
![image](https://github.com/PrismLauncher/PrismLauncher/assets/508861/b2951d03-a6d7-4174-a1b4-04a503d76c7a)

because `sponge-mixin` is uing a `+` in it's file name

we need to take a look at our library path code and at least ensure we are *consistent* with our replacement of bad path charaters

--- a/launcher/FileSystem.cpp
+++ b/launcher/FileSystem.cpp
@@ -831,14 +831,16 @@ QString RemoveInvalidPathChars(QString path, QChar replaceWith)
     invalidChars = BAD_WIN_CHARS;
 #endif
 
+    // the null character is ignored in this check as it was not a problem until now
     switch (statFS(path).fsType) {
         case FilesystemType::FAT:
             invalidChars += BAD_FAT_CHARS;
             break;
         case FilesystemType::NTFS:
+        /* fallthrough */
+        case FilesystemType::REFS:  // similar to NTFS(should be available only on windows)
             invalidChars += BAD_NTFS_CHARS;
             break;
-        // case FilesystemType::REFS:
         // case FilesystemType::EXT:
         // case FilesystemType::EXT_2_OLD:
         // case FilesystemType::EXT_2_3_4:

--- a/launcher/FileSystem.cpp
+++ b/launcher/FileSystem.cpp
@@ -74,6 +74,7 @@
 #include <objbase.h>
 #include <shlobj.h>
 #else
+#include <sys/statvfs.h>
 #include <utime.h>
 #endif
 
@@ -801,25 +802,97 @@ QString NormalizePath(QString path)
     }
 }
 
-static const QString BAD_PATH_CHARS = "\"?<>:;*|!\r\n";
-static const QString BAD_FILENAME_CHARS = BAD_PATH_CHARS + "\\/";
+QString removeDuplicates(QString a)
+{
+    auto b = a.split("");
+    b.removeDuplicates();
+    return b.join("");
+}
+
+QString getFileSystemType(const QString& path)
+{
+    QString fileSystemType;
+
+#ifdef Q_OS_WIN
+    wchar_t volume[MAX_PATH + 1] = { 0 };
+    if (GetVolumeInformationW((LPCWSTR)path.utf16(), nullptr, 0, nullptr, nullptr, nullptr, volume, MAX_PATH)) {
+        fileSystemType = QString::fromWCharArray(volume);
+    }
+#elif defined(Q_OS_UNIX)
+    struct statvfs buf;
+    if (statvfs(path.toUtf8().constData(), &buf) == 0) {
+        switch (buf.f_type) {
+            case 0x4d44:  // "MSDOS"
+                fileSystemType = "FAT32";
+                break;
+            case 0x5346544e:  // "NTFS"
+                fileSystemType = "NTFS";
+                break;
+            case 0x4244:  // "HFS+" or "H+" on some systems
+            case 0x482b:  // "HFS+" or "H+" on some systems
+                fileSystemType = "HFS+";
+                break;
+            case 0x41465342:  // "APFS"
+                fileSystemType = "APFS";
+                break;
+            case 0x65735546:  // "exFAT"
+                fileSystemType = "exFAT";
+                break;
+            default:
+                break;
+        }
+    }
+#endif
+
+    return fileSystemType;
+}
+
+static const QString BAD_WIN_CHARS = "\"?<>:*|\r\n";
+
+static const QString BAD_FAT32_CHARS = "<>:\"|?*+.,;=[]!";
+static const QString BAD_NTFS_CHARS = "<>:\"|?*";
+static const QString BAD_HFS_CHARS = ":";
+static const QString BAD_EXFAT_CHARS = "<>:\"|?*";
+
+static const QString BAD_FILENAME_CHARS =
+    removeDuplicates(BAD_WIN_CHARS + BAD_FAT32_CHARS + BAD_NTFS_CHARS + BAD_HFS_CHARS + BAD_EXFAT_CHARS) + "\\/";
 
 QString RemoveInvalidFilenameChars(QString string, QChar replaceWith)
 {
     for (int i = 0; i < string.length(); i++)
         if (string.at(i) < ' ' || BAD_FILENAME_CHARS.contains(string.at(i)))
             string[i] = replaceWith;
-
     return string;
 }
 
-QString RemoveInvalidPathChars(QString string, QChar replaceWith)
+QString RemoveInvalidPathChars(QString path, QChar replaceWith)
 {
-    for (int i = 0; i < string.length(); i++)
-        if (string.at(i) < ' ' || BAD_PATH_CHARS.contains(string.at(i)))
-            string[i] = replaceWith;
+    QString invalidChars;
+#ifdef Q_OS_WIN
+    invalidChars = BAD_WIN_CHARS;
+#endif
 
-    return string;
+    QString fileSystemType = getFileSystemType(QFileInfo(path).absolutePath());
+
+    if (fileSystemType == "FAT32") {
+        invalidChars += BAD_FAT32_CHARS;
+    } else if (fileSystemType == "NTFS") {
+        invalidChars += BAD_NTFS_CHARS;
+    } else if (fileSystemType == "HFS+" || fileSystemType == "APFS") {
+        invalidChars += BAD_HFS_CHARS;
+    } else if (fileSystemType == "exFAT") {
+        invalidChars += BAD_EXFAT_CHARS;
+    }
+
+    if (invalidChars.size() != 0) {
+        for (int i = 0; i < path.length(); i++) {
+            if (path.at(i) < ' ' || invalidChars.contains(path.at(i))) {
+                path[i] = replaceWith;
+            }
+        }
+    }
+
+    return path;
 }
 
 QString DirNameFromString(QString string, QString inDir)

--- a/launcher/FileSystem.cpp
+++ b/launcher/FileSystem.cpp
@@ -74,7 +74,6 @@
 #include <objbase.h>
 #include <shlobj.h>
 #else
-#include <sys/statvfs.h>
 #include <utime.h>
 #endif
 
@@ -809,53 +808,13 @@ QString removeDuplicates(QString a)
     return b.join("");
 }
 
-QString getFileSystemType(const QString& path)
-{
-    QString fileSystemType;
-
-#ifdef Q_OS_WIN
-    wchar_t volume[MAX_PATH + 1] = { 0 };
-    if (GetVolumeInformationW((LPCWSTR)path.utf16(), nullptr, 0, nullptr, nullptr, nullptr, volume, MAX_PATH)) {
-        fileSystemType = QString::fromWCharArray(volume);
-    }
-#elif defined(Q_OS_UNIX)
-    struct statvfs buf;
-    if (statvfs(path.toUtf8().constData(), &buf) == 0) {
-        switch (buf.f_type) {
-            case 0x4d44:  // "MSDOS"
-                fileSystemType = "FAT32";
-                break;
-            case 0x5346544e:  // "NTFS"
-                fileSystemType = "NTFS";
-                break;
-            case 0x4244:  // "HFS+" or "H+" on some systems
-            case 0x482b:  // "HFS+" or "H+" on some systems
-                fileSystemType = "HFS+";
-                break;
-            case 0x41465342:  // "APFS"
-                fileSystemType = "APFS";
-                break;
-            case 0x65735546:  // "exFAT"
-                fileSystemType = "exFAT";
-                break;
-            default:
-                break;
-        }
-    }
-#endif
-
-    return fileSystemType;
-}
-
 static const QString BAD_WIN_CHARS = "\"?<>:*|\r\n";
 
-static const QString BAD_FAT32_CHARS = "<>:\"|?*+.,;=[]!";
+static const QString BAD_FAT_CHARS = "<>:\"|?*+.,;=[]!";
 static const QString BAD_NTFS_CHARS = "<>:\"|?*";
 static const QString BAD_HFS_CHARS = ":";
-static const QString BAD_EXFAT_CHARS = "<>:\"|?*";
 
-static const QString BAD_FILENAME_CHARS =
-    removeDuplicates(BAD_WIN_CHARS + BAD_FAT32_CHARS + BAD_NTFS_CHARS + BAD_HFS_CHARS + BAD_EXFAT_CHARS) + "\\/";
+static const QString BAD_FILENAME_CHARS = removeDuplicates(BAD_WIN_CHARS + BAD_FAT_CHARS + BAD_NTFS_CHARS + BAD_HFS_CHARS) + "\\/";
 
 QString RemoveInvalidFilenameChars(QString string, QChar replaceWith)
 {
@@ -872,16 +831,35 @@ QString RemoveInvalidPathChars(QString path, QChar replaceWith)
     invalidChars = BAD_WIN_CHARS;
 #endif
 
-    QString fileSystemType = getFileSystemType(QFileInfo(path).absolutePath());
-
-    if (fileSystemType == "FAT32") {
-        invalidChars += BAD_FAT32_CHARS;
-    } else if (fileSystemType == "NTFS") {
-        invalidChars += BAD_NTFS_CHARS;
-    } else if (fileSystemType == "HFS+" || fileSystemType == "APFS") {
-        invalidChars += BAD_HFS_CHARS;
-    } else if (fileSystemType == "exFAT") {
-        invalidChars += BAD_EXFAT_CHARS;
+    switch (statFS(path).fsType) {
+        case FilesystemType::FAT:
+            invalidChars += BAD_FAT_CHARS;
+            break;
+        case FilesystemType::NTFS:
+            invalidChars += BAD_NTFS_CHARS;
+            break;
+        // case FilesystemType::REFS:
+        // case FilesystemType::EXT:
+        // case FilesystemType::EXT_2_OLD:
+        // case FilesystemType::EXT_2_3_4:
+        // case FilesystemType::XFS:
+        // case FilesystemType::BTRFS:
+        // case FilesystemType::NFS:
+        // case FilesystemType::ZFS:
+        case FilesystemType::APFS:
+        /* fallthrough */
+        case FilesystemType::HFS:
+        /* fallthrough */
+        case FilesystemType::HFSPLUS:
+        /* fallthrough */
+        case FilesystemType::HFSX:
+            invalidChars += BAD_HFS_CHARS;
+            break;
+        // case FilesystemType::FUSEBLK:
+        // case FilesystemType::F2FS:
+        // case FilesystemType::UNKNOWN:
+        default:
+            break;
     }
 
     if (invalidChars.size() != 0) {

--- a/launcher/MMCZip.cpp
+++ b/launcher/MMCZip.cpp
@@ -288,9 +288,7 @@ std::optional<QStringList> extractSubDir(QuaZip* zip, const QString& subdir, con
 
     do {
         QString file_name = zip->getCurrentFileName();
-#ifdef Q_OS_WIN
         file_name = FS::RemoveInvalidPathChars(file_name);
-#endif
         if (!file_name.startsWith(subdir))
             continue;
 

--- a/launcher/minecraft/Library.cpp
+++ b/launcher/minecraft/Library.cpp
@@ -51,6 +51,7 @@ void Library::getApplicableFiles(const RuntimeContext& runtimeContext,
 {
     bool local = isLocal();
     auto actualPath = [&](QString relPath) {
+        relPath = FS::RemoveInvalidPathChars(relPath);
         QFileInfo out(FS::PathCombine(storagePrefix(), relPath));
         if (local && !overridePath.isEmpty()) {
             QString fileName = out.fileName();

--- a/launcher/modplatform/flame/FlameInstanceCreationTask.cpp
+++ b/launcher/modplatform/flame/FlameInstanceCreationTask.cpp
@@ -538,9 +538,7 @@ void FlameCreationTask::setupDownloadJob(QEventLoop& loop)
     }
     for (const auto& result : results) {
         auto fileName = result.fileName;
-#ifdef Q_OS_WIN
         fileName = FS::RemoveInvalidPathChars(fileName);
-#endif
         auto relpath = FS::PathCombine(result.targetFolder, fileName);
 
         if (!result.required && !selectedOptionalMods.contains(relpath)) {

--- a/launcher/modplatform/flame/FlameModIndex.cpp
+++ b/launcher/modplatform/flame/FlameModIndex.cpp
@@ -139,9 +139,7 @@ auto FlameMod::loadIndexedPackVersion(QJsonObject& obj, bool load_changelog) -> 
     file.version = Json::requireString(obj, "displayName");
     file.downloadUrl = Json::ensureString(obj, "downloadUrl");
     file.fileName = Json::requireString(obj, "fileName");
-#ifdef Q_OS_WIN
     file.fileName = FS::RemoveInvalidPathChars(file.fileName);
-#endif
 
     ModPlatform::IndexedVersionType::VersionType ver_type;
     switch (Json::requireInteger(obj, "releaseType")) {

--- a/launcher/modplatform/modrinth/ModrinthInstanceCreationTask.cpp
+++ b/launcher/modplatform/modrinth/ModrinthInstanceCreationTask.cpp
@@ -241,9 +241,7 @@ bool ModrinthCreationTask::createInstance()
 
     for (auto file : m_files) {
         auto fileName = file.path;
-#ifdef Q_OS_WIN
         fileName = FS::RemoveInvalidPathChars(fileName);
-#endif
         auto file_path = FS::PathCombine(root_modpack_path, fileName);
         if (!root_modpack_url.isParentOf(QUrl::fromLocalFile(file_path))) {
             // This means we somehow got out of the root folder, so abort here to prevent exploits

--- a/launcher/modplatform/modrinth/ModrinthPackIndex.cpp
+++ b/launcher/modplatform/modrinth/ModrinthPackIndex.cpp
@@ -227,9 +227,7 @@ auto Modrinth::loadIndexedPackVersion(QJsonObject& obj, QString preferred_hash_t
     if (parent.contains("url")) {
         file.downloadUrl = Json::requireString(parent, "url");
         file.fileName = Json::requireString(parent, "filename");
-#ifdef Q_OS_WIN
         file.fileName = FS::RemoveInvalidPathChars(file.fileName);
-#endif
         file.is_preferred = Json::requireBoolean(parent, "primary") || (files.count() == 1);
         auto hash_list = Json::requireObject(parent, "hashes");
 

--- a/launcher/net/HttpMetaCache.cpp
+++ b/launcher/net/HttpMetaCache.cpp
@@ -84,9 +84,7 @@ auto HttpMetaCache::getEntry(QString base, QString resource_path) -> MetaEntryPt
 
 auto HttpMetaCache::resolveEntry(QString base, QString resource_path, QString expected_etag) -> MetaEntryPtr
 {
-#ifdef Q_OS_WIN
     resource_path = FS::RemoveInvalidPathChars(resource_path);
-#endif
     auto entry = getEntry(base, resource_path);
     // it's not present? generate a default stale entry
     if (!entry) {

